### ISSUE TITLE
kubectl logs should allow to discriminate between pod's stderr and stdout

### ISSUE THREAD

**Issue by pcm32** [kubernetes/kubernetes#28167]
In the current state, invocation of `kubectl logs pods/my-pod` provides mixed stderr and stdout from the pod. For certain applications, it is important to be able to separate stderr from stdout. I haven't found any way of separating stderr from stdout through `kubectl logs...`.

**Comment by sakshamsharma** [comment]
I had once had a discussion on this topic with a friend, unrelated to k8s. One simple solution was to pipe both into different files. But that left the issue that it was hard to combine them in-order.

I'm trying to look at the k8s code to see how it is done here. In any case, if the team appreciates this, I could work on it after a discussion.

**Comment by foxish** [comment]
@kubernetes/sig-node-feature-requests This seems rather old. Was this abandoned for some reason? It seems nifty for some applications (like Spark) which have useful stdout and stderr streams that we'd want to access but separately since stderr is quite noisy.

**Comment by ncdc** [comment]
Separate stdout and stderr log streams are supported in the kubelet. They are not, however, exposed from the kubelet's http server upward. See [here](https://github.com/kubernetes/kubernetes/blob/b011529d8a1486bc2316a049db35759086d2994b/pkg/kubelet/server/server.go#L507-L515) where the same writer is passed for both stdout and stderr (the last 2 params to `s.host.GetKubeletContainerLogs`).

I'm guessing we would need to define a protocol for streaming back separate streams for logs, as the current implementation just does a straight `io.Copy` [from the http response to stdout](https://github.com/kubernetes/kubernetes/blob/5472a5e0a591d49c14c28a69ed97200d7c3eacae/pkg/kubectl/cmd/logs.go#L257).

cc @smarterclayton @sttts

**Comment by ncdc** [comment]
cc @derekwaynecarr @kubernetes/rh-cluster-infra

**Comment by pcm32** [comment]
Thanks @ncdc and others for looking into this! so there is a solution in the horizon then! Our use case is related to scientific workflows, where we have the ability to run a workflow environment (Galaxy, used in Bioinformatics) inside Kubernetes and offloads jobs from that environment to the same Kubernetes cluster. Currently we just our all stderr/stdout inside a single file as it is what the logs command allows, but this sort of breaks what the user expects as output in the workflow environment. So it would be great to have the functionality!

**Comment by ironhouzi** [comment]
Just wanted to add that there are others in need of this feature for the exact same reasons as @pcm32 describes. I've tried finding som hacky workarounds, but it gets really ugly very quickly.. Would be great to have this feature implemented.

**Comment by jwendell** [comment]
I'm working on this, have a working implementation. It's only missing the tests.

What I did was to add another field to v1.PodLogOptions, indicating which stream the user wants: stdout, stderr or both (default if not provided, to keep current behavior). This is easy to add/implement as it doesn't break API.

@ncdc , by "define a protocol for streaming back separate streams for logs" you mean return a proper json response instead of plain text as of today? It would break API, no?

**Comment by ncdc** [comment]
Yes, API breakage. Could be json, could be something else. Could maybe
change what's returned based on client version?

**Comment by smarterclayton** [comment]
If you want separate streams the websocket endpoint works straight up

**Comment by smarterclayton** [comment]
We can switch kubectl logs to using web sockets and only fall back if the server rejects it (web socket log streams have been in since 1.3 or so)

**Comment by atombender** [comment]
I propose that a new API is introduced to fix this. One of problems  issues with the current API is that errors from Kubernetes itself cannot be distinguished from errors originating from the container; it's all just text. For example, the `unexpected stream type ""` error (see #47800), which comes from Kubernetes, is an out-of-band error message that is intermingled with the output from the container. Separating stdout/stderr doesn't help here, because the Kubernetes error is neither. 

A better approach would be define a new API, or use content header negotation (e.g. `Accept: application/json`), that wrap the entries in JSON. Something like NDJSON, perhaps:

```
{"stream":"stdout","log":"hello world","t":"2017-08-01T17:14:48+02:00"}
{"stream":"error","log":"unexpected stream type \"\"","t":"2017-08-01T17:14:48+02:00"}
{"stream":"stderr","log":"some application error","t":"2017-08-01T17:14:48+02:00"}
```

**Comment by txomon** [comment]
Is there any update or progress on this?

**Comment by javadnoorb** [comment]
Any updates on this?

**Comment by codemug** [comment]
Any updates on this?

**Comment by lewisgcm** [comment]
@smarterclayton is there any documentation regarding the logs websocket endpoint and getting separate streams?

I found this https://gitlab.cncf.ci/kubernetes/kubernetes/commit/363b61690839b5af92c846c632e7afb0771353bb although using the `channel.k8s.io` WebSocket protocol with the logs endpoint returns a 403, the default `binary.k8s.io` does work but stdout/stderr are merged.

**Comment by gWOLF3** [comment]
@kubernetes/sig-node-feature-requests

**Comment by nojvek** [comment]
Is this something the kubernetes authors are interested in accepting. I'd be interested to land a PR for this if someone is able to guide me. 

I assume the logs endpoint needs to take a flag and pipe this down all the way to the place that's getting the logs from the various containers.

**Comment by dims** [comment]
/cc

long-term-issue (note to self)

**Comment by zhixiangjoy** [comment]
Any updates on this?

**Comment by RichardHill** [comment]
We are currently working on a software solution that would very much benefit from having this resolved. It seems very strange to me that these two should end up being merged together and I'm left wondering what the engineering decision was to come the conclusion that it would be a good idea!!

**Comment by m4nu56** [comment]
We also would greatly benefit from having the stdout and stderr separated when reading the logs with `kubectl logs ..`. Is there any chance this could be resolved in the future ? 
Do you guys have any alternative solution that doesn't require to run sidecar containers along to collect the stderr of the aimed container ?

**Comment by knight42** [comment]
I would like to work on this.
/assign

There seems to be lots of work to do, I will draft a KEP first, then we could discuss about the details.

**Comment by knight42** [comment]
Hi all! For those who are willing to see this feature get implemented, I have submited my proposal https://github.com/kubernetes/enhancements/pull/3289, please help to review the design details. Once the proposal is merged, I am able to get to work on it! Thanks in advance!

**Comment by Xernisfy** [comment]
Is this done, since https://github.com/kubernetes/enhancements/pull/3289 was merged?

**Comment by zwpaper** [comment]
@Xernisfy nope, only KEP get merged, it still need to be implemented

**Comment by eero-t** [comment]
> Hi all! For those who are willing to see this feature get implemented, I have submited my proposal [kubernetes/enhancements#3289](https://github.com/kubernetes/enhancements/pull/3289), please help to review the design details.

@knight42, @atombender suggested above (in 2017) separating also k8s errors from the mixed stream.  Is that not any more an issue?
